include/sleipnir/control/OCPSolver.hpp

```diff
@@ -0,0 +1,257 @@
+   * @param timestepMethod The timestep method.
+   * @param method The transcription method.
+   */
+  OCPSolver(
+      int numStates, int numInputs, std::chrono::duration<double> dt,
+      int numSteps, const DynamicsFunction& dynamics,
+      DynamicsType dynamicsType = DynamicsType::kExplicitODE,
+      TimestepMethod timestepMethod = TimestepMethod::kFixed,
+      TranscriptionMethod method = TranscriptionMethod::kDirectTranscription);
+
+  /**
+   * Utility function to constrain the initial state.
+   *
+   * @param initialState the initial state to constrain to.
+   */
+  void ConstrainInitialState(const VariableMatrix& initialState);
+
+  /**
+   * Utility function to constrain the final state.
+   *
+   * @param finalState the final state to constrain to.
+   */
+
+  void ConstrainFinalState(const VariableMatrix& finalState);
+
+  /**
+   * Set the constraint evaluation function. This function is called
+   * `numSteps+1` times, with the corresponding state and input
+   * VariableMatrices.
+   *
+   * @param constraintFunction the constraint function.
+   */
+  void ConstrainAlways(const OCPConstraintCallback& constraintFunction);
+
+  /**
+   * Convenience function to set a lower bound on the input.
+   *
+   * @param lowerBound The lower bound that inputs must always be above. Must be
+   * shaped (numInputs)x1.
+   */
+  void SetLowerInputBound(const VariableMatrix& lowerBound) {
+    for (int i = 0; i < m_numSteps + 1; ++i) {
+      SubjectTo(U().Col(i) >= lowerBound);
+    }
+  }
+
+  /**
+   * Convenience function to set an upper bound on the input.
+   *
+   * @param upperBound The upper bound that inputs must always be below. Must be
+   * shaped (numInputs)x1.
+   */
+  void SetUpperInputBound(const VariableMatrix& upperBound) {
+    for (int i = 0; i < m_numSteps + 1; ++i) {
+      SubjectTo(U().Col(i) <= upperBound);
+    }
+  }
+
+  /**
+   * Convenience function to set an upper bound on the timestep.
+   *
+   * @param maxTimestep The maximum timestep.
+   */
+  void SetMaxTimestep(std::chrono::duration<double> maxTimestep) {
+    SubjectTo(DT() <= maxTimestep.count());
+  }
+
+  /**
+   * Convenience function to set a lower bound on the timestep.
+   *
+   * @param minTimestep The minimum timestep.
+   */
+  void SetMinTimestep(std::chrono::duration<double> minTimestep) {
+    SubjectTo(DT() >= minTimestep.count());
+  }
+
+  /**
+   * Get the state variables. After the problem is solved, this will contain the
+   * optimized trajectory.
+   *
+   * Shaped (numStates)x(numSteps+1).
+   *
+   * @returns The state variable matrix.
+   */
+  VariableMatrix& X() { return m_X; };
+
+  /**
+   * Get the input variables. After the problem is solved, this will contain the
+   * inputs corresponding to the optimized trajectory.
+   *
+   * Shaped (numInputs)x(numSteps+1), although the last input step is unused in
+   * the trajectory.
+   *
+   * @returns The input variable matrix.
+   */
+  VariableMatrix& U() { return m_U; };
```

Comment:
Should these getters be const-qualified and return const refs? I had trouble with types and templating, especially re: integration when trying to switch them over so I quit, and I'm not sure if it would significantly improve the performance.